### DISCUSSION TITLE
Programmatically registering self-hosted Runners

### DISCUSSION THREAD

**Discussion by jmaukisch** [community/community#27204] [Actions]
<p>Hi all,</p>
<p>i’d like to programmatically register self-hosted Actions Runner on a given repository, but i have some issues obtaining the registration token that is needed to invoke the config.sh.</p>
<p>What i got so far:</p>
<ul>
<li>I have an GitHub App installed on the Repository</li>
<li>I gerate a JWT</li>
<li>I call <a href="http://api.github.com/app/installations/123456/access_tokens" rel="noopener nofollow ugc">api.github.com/app/installations/123456/access_tokens</a> with the JWT to get an access token for that repo for the installation</li>
<li>and with the returned token I try to call <a href="http://api.github.com/repos/owner/my-repo/actions/runners/registration-token" rel="noopener nofollow ugc">api.github.com/repos/owner/my-repo/actions/runners/registration-token</a> for the repository</li>
</ul>
<p>But i always end up with  the response “Resource not accessible by integration”.<br>
Is it not possible to access this endpoint with an installation access token or is my app missing some permissions for that?<br>
Or did i miss some steps on the way?</p>
<p>Thanks for your help &amp; kind regards,<br>
Jens</p>

**Comment by michaelsync** [comment]
<p>Have you tried using Personal access token?</p>
<p>I am registering the self-hosted runner from Dockerfile programmatically. Because I want to run the self-hosted runner inside the docker container. I got it working with PAT… Not with Github app.</p>
<p>I think if you are using the Gihub app, you could probably check the permission in the document. I think it needs to have the full right for repo and self-host permission. For organization runner, u need to give the full admin right too.</p>

**Comment by jmaukisch** [comment]
<p>Thanks, we are currently doing in the same way as you do with the PAT, but we rather want to do it via a GitHub app for some different reasons.</p>
<p>And i just found out what i was missing to get i work with the GitHub App:<br>
somehow i forgot to give it the required administration permission (<a href="https://docs.github.com/en/free-pro-team@latest/rest/reference/permissions-required-for-github-apps#permission-on-administration" rel="noopener nofollow ugc">https://docs.github.com/en/free-pro-team@latest/rest/reference/permissions-required-for-github-apps#permission-on-administration</a>) so that endpoint can be used by the app.</p>
<p>With that everything works fine :slight_smile:</p>
<p>kind regards,<br>
jens</p>

**Comment by michaelsync** [comment]
<p>Thanks. That’s a great new.</p>
<p>Can you please share what you did with Github app?</p>
<p>I did the followings.</p>
<ul>
<li>Enter Homepage URL</li>
<li>No User authorization callback URL</li>
<li>Uncheck “Expire user authorization tokens”</li>
<li>Uncheck Webhook Active</li>
<li>Organization Permission (Admin: R/W, Self-hosted runners: R/W)</li>
<li>Generate the private key and download pem</li>
</ul>
<p>I don’t like using the PAT. I’ve been asking around as below… No luck yet.</p>
<ul>
<li><a href="https://github.community/t/can-we-use-the-github-app-or-oauth-app-instead-of-github-pat-to-generate-the-runner-token/146505" class="inline-onebox">Can we use the github app or oauth app instead of github PAT to generate the RUNNER_TOKEN?</a></li>
<li><a href="https://stackoverflow.com/questions/65090500/how-to-use-the-github-app-oauth-app-to-register-the-github-action-self-hosted-ru" rel="noopener nofollow ugc">https://stackoverflow.com/questions/65090500/how-to-use-the-github-app-oauth-app-to-register-the-github-action-self-hosted-ru</a></li>
</ul>

**Comment by jmaukisch** [comment, marked as answer]
<p>What I did so far is:</p>
<ul>
<li>Created an app with the " Repository permissions" “Administration: Read &amp; Write”</li>
<li>Generate and Download the Private Key</li>
<li>Install the App on a Repository</li>
<li>Used the Ruby Script form the <a href="https://docs.github.com/en/free-pro-team@latest/developers/apps/authenticating-with-github-apps" rel="noopener nofollow ugc">Documentation</a> to generate a JWT for the App</li>
<li>Used the generated JWT to call the API to get an Application Access Token:</li>
</ul>
<pre><code class="lang-auto">curl -i -X POST \
-H "Authorization: Bearer &lt;JWT&gt;" \
-H "Accept: application/vnd.github.v3+json" \
https://api.github.com/app/installations/&lt;Installation Id of your App for the Repo&gt;/access_tokens \
 -d '{"repositories":["name-of-your-repo"]}'
</code></pre>
<ul>
<li>Use the token in the response to call the endpoint to get the registration token for the runner for the repo</li>
</ul>
<pre><code class="lang-auto">curl \
  -X POST \
  -H "Authorization: token &lt;token&gt;" \
  -H "Accept: application/vnd.github.machine-man-preview+json" \
  https://api.github.com/repos/&lt;owner&gt;/&lt;name-of-your-repo&gt;/actions/runners/registration-token
</code></pre>
<ul>
<li>use the token from the response to call the <code>./config.sh</code> from the runner</li>
</ul>
<p>So this is all on a repository level, I did not try it for an organisation or so.</p>

**Comment by DrMerlin** [comment]
<p>I found I could do it with a personal access token as long as it had:</p>
<ul>
<li>repo</li>
<li>admin:repo_hook</li>
</ul>
<p>perms.  Then it was just the one step with the personal access token.</p>

**Comment by larryclaman** [comment]
<p>Interesting… I’ve been looking for this info but could not find it.  Is this (eg how to use a PAT with a self-hosted runner) officially documented anywhere?</p>
